framework/src/outputs/PNGOutput.C

```diff
@@ -0,0 +1,228 @@
+  return params;
+}
+
+
+PNGOutput::PNGOutput(const InputParameters & parameters) :
+Output(parameters),
+resolution(getParam<Real>("resolution")),
+PNGFile(getParam<std::string>("PNGFile")),
+testStepToPNG(getParam<Real>("testStepToPNG")),
+inColor(getParam<bool>("inColor")),
+outBoundsControlColor(getParam<Real>("outBoundsControlColor"))
+{
+}
+
+// Funtion for making the _mesh_function object.
+void PNGOutput::makeMeshFunc() {
+
+  const std::vector<unsigned int> var_nums = {0};
+
+  // Find the values that will be used for rescaling purposes.
+  calculateRescalingValues();
+
+  // Set up the mesh_function
+  _mesh_function = libmesh_make_unique<MeshFunction>(
+      *_es_ptr, nls.serializedSolution(), nls.dofMap(), var_nums);
+  _mesh_function->init();
+
+  // Need to enable out of mesh with the given control color scaled in reverse
+  // scaling is done, this value retains it's original value.
+  _mesh_function->enable_out_of_mesh_mode(reverseScale);
+}
+
+
+void PNGOutput::calculateRescalingValues()
+{
+  // The min and max.
+  scalingMin = nls.serializedSolution().min();
+  scalingMax = nls.serializedSolution().max();
+  shiftValue = 0;
+
+  // Get the shift value.
+  if(scalingMin != 0)
+  {
+    // Shiftvalue will be the same magnitude, but
+    // going in the opposite direction of the scalingMin
+    shiftValue -= scalingMin;
+  }
+
+  // Shift the max.
+  scalingMax += shiftValue;
+
+}
+
+
+void PNGOutput::setRGB(png_byte *rgb, Real selection) {
+  // Using an RGB system with Red as 0 - 255, Green as 256 - 511 and
+  // Blue as 512 - 767 gives us our total colorSpectrum of 0 - 767.
+  const Real colorSpectrumMax = 767;
+  // We need to convert the
+  auto color = (int)(selection * colorSpectrumMax);
+  // Make sure everything is within our colorSpectrum.
+  if(color > colorSpectrumMax)
+    color = colorSpectrumMax;
+  if(color < 0)
+    color = 0;
+  Real magnitude = color % 256;
+
+  // Current color scheme: Blue->Red->Yellow->White
+  // Blue->Red
+  if(color < 256)
+  {
+    rgb[0] = magnitude;
+    rgb[1] = 0;
+    rgb[2] = 255-magnitude;
+  }
+  // Red->Yellow
+  else if(color < 512)
+  {
+    rgb[0] = 255;
+    rgb[1] = magnitude;
+    rgb[2] = 0;
+  }
+  // Yellow->White
+  else
+  {
+    rgb[0] = 255;
+    rgb[1] = 255;
+    rgb[2] = magnitude;
+  }
+}
+
+
+void PNGOutput::output(const ExecFlagType & /*type*/)
+{
+  makeMeshFunc();
+  box = MeshTools::create_bounding_box(*_mesh_ptr);
+  makePNG();
+}
+
+void PNGOutput::makePNG() {
+  // Increment the testStep
+  testStep++;
+
+  // Get the max and min of the BoundingBox
+  Point maxPoint = box.max();
+  Point minPoint = box.min();
+
+  // The the total distance on the x and y axes.
+  Real distx = maxPoint(0) - minPoint(0);
+  Real disty = maxPoint(1) - minPoint(1);
+
+  // Create the filename based on base and the test step number.
+  std::string PNGFile2 = PNGFile;
+  PNGFile2 += std::to_string((int)testStep);
```

Comment:
Good - however, we generally like padding so that sorting works better. We typically do `foo_001.png foo_002.png, etc`. You can do this with `std::setw` and `std::fill`.

Reply:
This was confusing me a bit, but I think I can get it once I find the right way to use _file_base.